tests/test_help.py

```diff
@@ -390,36 +386,79 @@ def test_tool_exec_full_help(sysargv, caplog):
         "-v, --verbose",
     ]
 
-    # check the result of the full help builder is what is shown
-    expected = "test help"
-    assert [expected] == [rec.message for rec in caplog.records]
+
+def test_tool_exec_command_incorrect(help_builder):
+    """Execute a command that doesn't exist."""
+    help_builder.init("testapp", "general summary", [])
+    with pytest.raises(ArgumentParsingError) as cm:
+        Dispatcher(["wrongcommand"], [])
+
+    expected = textwrap.dedent(
+        """\
+        Usage: testapp [options] command [args]...
+        Try 'testapp -h' for help.
+
+        Error: no such command 'wrongcommand'
+        """
+    )
+
+    error = cm.value
+    assert str(error) == expected
 
 
 @pytest.mark.parametrize(
     "sysargv",
     [
-        ["wrongcommand"],
         ["-h", "wrongcommand"],
         ["wrongcommand", "-h"],
         ["--help", "wrongcommand"],
         ["wrongcommand", "--help"],
         ["-h", "wrongcommand", "--help"],
     ],
 )
-def test_tool_exec_command_incorrect(sysargv, help_builder):
+def test_tool_exec_help_on_command_incorrect(sysargv, help_builder):
     """Execute a command that doesn't exist."""
-    command_groups = [("group", "help", [HelpCommand])]  # the only command we need
-    help_builder.init("testapp", "general summary", command_groups)
+    help_builder.init("testapp", "general summary", [])
     with pytest.raises(ArgumentParsingError) as cm:
-        dispatcher = Dispatcher(sysargv, command_groups)
-        dispatcher.run()
+        Dispatcher(sysargv, [])
 
     expected = textwrap.dedent(
         """\
         Usage: testapp [options] command [args]...
         Try 'testapp -h' for help.
 
-        Error: no such command 'wrongcommand'
+        Error: command 'wrongcommand' not found to provide help for
+        """
+    )
+
+    error = cm.value
+    assert str(error) == expected
+
+
+@pytest.mark.parametrize(
+    "sysargv",
+    [
+        ["-h", "foo", "bar"],
+        ["foo", "-h", "bar"],
+        ["foo", "bar", "-h"],
+        ["--help", "foo", "bar"],
+        ["foo", "--help", "bar"],
+        ["foo", "bar", "--help"],
+        ["help", "foo", "bar"],
+    ],
+)
+def test_tool_exec_help_on_too_many_things(sysargv, help_builder):
+    """Trying to get help on too many items."""
+    help_builder.init("testapp", "general summary", [])
+    with pytest.raises(ArgumentParsingError) as cm:
+        Dispatcher(sysargv, [])
+
+    expected = textwrap.dedent(
+        """\
+        Usage: testapp [options] command [args]...
+        Try 'testapp -h' for help.
+
+        Error: Too many parameters when requesting help; pass a command, '--all', or leave it empty
```

Comment:
I am not fan of "leave it empty" but I cannot find other words for it

Reply:
I had the same struggle. In 2nd place came "or nothing."